File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/amp/CachingCatalog.java
/**
 * @file
 * @brief Caching catalog implementation
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.amp;

import java.io.IOException;

import app.zxtune.TimeStamp;
import app.zxtune.fs.dbhelpers.CommandExecutor;
import app.zxtune.fs.dbhelpers.QueryCommand;
import app.zxtune.fs.dbhelpers.Timestamps;

final public class CachingCatalog implements Catalog {

  private final TimeStamp GROUPS_TTL = TimeStamp.fromDays(30);
  private final TimeStamp AUTHORS_TTL = TimeStamp.fromDays(30);
  private final TimeStamp TRACKS_TTL = TimeStamp.fromDays(30);

  private final RemoteCatalog remote;
  private final Database db;
  private final CommandExecutor executor;

  CachingCatalog(RemoteCatalog remote, Database db) {
    this.remote = remote;
    this.db = db;
    this.executor = new CommandExecutor("amp");
  }

  @Override
  public void queryGroups(final GroupsVisitor visitor) throws IOException {
    executor.executeQuery("groups", new QueryCommand() {

      private final Timestamps.Lifetime lifetime = db.getGroupsLifetime(GROUPS_TTL);

      @Override
      public boolean isCacheExpired() {
        return lifetime.isExpired();
      }

      @Override
      public void updateCache() throws IOException {
        db.runInTransaction(() -> {
          remote.queryGroups(new GroupsVisitor() {
            // TODO: remove
            @Override
            public void setCountHint(int count) {}

            @Override
            public void accept(Group obj) {
              db.addGroup(obj);
            }
          });
          lifetime.update();
        });
      }

      @Override
      public boolean queryFromCache() {
        return db.queryGroups(visitor);
      }
    });
  }

  @Override
  public void queryAuthors(final String handleFilter, final AuthorsVisitor visitor) throws IOException {
    executor.executeQuery("authors", new QueryCommand() {

      private final Timestamps.Lifetime lifetime = db.getAuthorsLifetime(handleFilter, AUTHORS_TTL);

      @Override
      public boolean isCacheExpired() {
        return lifetime.isExpired();
      }

      @Override
      public void updateCache() throws IOException {
        db.runInTransaction(() -> {
          remote.queryAuthors(handleFilter, new AuthorsVisitor() {
            // TODO: remove
            @Override
            public void setCountHint(int count) {}

            @Override
            public void accept(Author obj) {
              db.addAuthor(obj);
            }
          });
          lifetime.update();
        });
      }

      @Override
      public boolean queryFromCache() {
        return db.queryAuthors(handleFilter, visitor);
      }
    });
  }

  @Override
  public void queryAuthors(final Country country, final AuthorsVisitor visitor) throws IOException {
    executor.executeQuery("authors", new QueryCommand() {

      private final Timestamps.Lifetime lifetime = db.getCountryLifetime(country, AUTHORS_TTL);

      @Override
      public boolean isCacheExpired() {
        return lifetime.isExpired();
      }

      @Override
      public void updateCache() throws IOException {
        db.runInTransaction(() -> {
          remote.queryAuthors(country, new AuthorsVisitor() {
            // TODO: remove
            @Override
            public void setCountHint(int count) {}

            @Override
            public void accept(Author obj) {
              db.addAuthor(obj);
              db.addCountryAuthor(country, obj);
            }
          });
          lifetime.update();
        });
      }

      @Override
      public boolean queryFromCache() {
        return db.queryAuthors(country, visitor);
      }
    });
  }

  @Override
  public void queryAuthors(final Group group, final AuthorsVisitor visitor) throws IOException {
    executor.executeQuery("authors", new QueryCommand() {

      private final Timestamps.Lifetime lifetime = db.getGroupLifetime(group, AUTHORS_TTL);

      @Override
      public boolean isCacheExpired() {
        return lifetime.isExpired();
      }

      @Override
      public void updateCache() throws IOException {
        db.runInTransaction(() -> {
          remote.queryAuthors(group, new AuthorsVisitor() {
            // TODO: remove
            @Override
            public void setCountHint(int count) {}

            @Override
            public void accept(Author obj) {
              db.addAuthor(obj);
              db.addGroupAuthor(group, obj);
            }
          });
          lifetime.update();
        });
      }

      @Override
      public boolean queryFromCache() {
        return db.queryAuthors(group, visitor);
      }
    });
  }

  @Override
  public void queryTracks(final Author author, final TracksVisitor visitor) throws IOException {
    executor.executeQuery("tracks", new QueryCommand() {

      private final Timestamps.Lifetime lifetime = db.getAuthorTracksLifetime(author, TRACKS_TTL);

      @Override
      public boolean isCacheExpired() {
        return lifetime.isExpired();
      }

      @Override
      public void updateCache() throws IOException {
        db.runInTransaction(() -> {
          remote.queryTracks(author, new TracksVisitor() {
            // TODO: remove
            @Override
            public void setCountHint(int count) {}

            @Override
            public void accept(Track obj) {
              db.addTrack(obj);
              db.addAuthorTrack(author, obj);
            }
          });
          lifetime.update();
        });
      }

      @Override
      public boolean queryFromCache() {
        return db.queryTracks(author, visitor);
      }
    });
  }

  @Override
  public void findTracks(String query, FoundTracksVisitor visitor) throws IOException {
    if (remote.searchSupported()) {
      remote.findTracks(query, visitor);
    } else {
      db.findTracks(query, visitor);
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/amp/CachingCatalog.kt
/**
 * @file
 * @brief Caching catalog implementation
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs.amp

import app.zxtune.TimeStamp
import app.zxtune.fs.dbhelpers.CommandExecutor
import app.zxtune.fs.dbhelpers.QueryCommand

private val GROUPS_TTL = TimeStamp.fromDays(30)
private val AUTHORS_TTL = TimeStamp.fromDays(30)
private val TRACKS_TTL = TimeStamp.fromDays(30)
private val PICTURES_TTL = TimeStamp.fromDays(30)

class CachingCatalog internal constructor(
    private val remote: RemoteCatalog,
    private val db: Database
) : Catalog {
    private val executor = CommandExecutor("amp")

    override fun queryGroups(visitor: Catalog.GroupsVisitor) =
        executor.executeQuery("groups", object : QueryCommand {
            private val lifetime = db.getGroupsLifetime(GROUPS_TTL)
            override val isCacheExpired
                get() = lifetime.isExpired

            override fun updateCache() = db.runInTransaction {
                remote.queryGroups(db::addGroup)
                lifetime.update()
            }

            override fun queryFromCache() = db.queryGroups(visitor)
        })

    override fun queryAuthors(handleFilter: String, visitor: Catalog.AuthorsVisitor) =
        executor.executeQuery("authors", object : QueryCommand {
            private val lifetime = db.getAuthorsLifetime(handleFilter, AUTHORS_TTL)
            override val isCacheExpired
                get() = lifetime.isExpired

            override fun updateCache() = db.runInTransaction {
                remote.queryAuthors(handleFilter, db::addAuthor)
                lifetime.update()
            }


            override fun queryFromCache() = db.queryAuthors(handleFilter, visitor)
        })

    override fun queryAuthors(country: Country, visitor: Catalog.AuthorsVisitor) =
        executor.executeQuery("authors", object : QueryCommand {
            private val lifetime = db.getCountryLifetime(country, AUTHORS_TTL)
            override val isCacheExpired
                get() = lifetime.isExpired

            override fun updateCache() {
                db.runInTransaction {
                    remote.queryAuthors(country) {
                        db.addAuthor(it)
                        db.addCountryAuthor(country, it)
                    }
                    lifetime.update()
                }
            }

            override fun queryFromCache() = db.queryAuthors(country, visitor)
        })

    override fun queryAuthors(group: Group, visitor: Catalog.AuthorsVisitor) =
        executor.executeQuery("authors", object : QueryCommand {
            private val lifetime = db.getGroupLifetime(group, AUTHORS_TTL)
            override val isCacheExpired
                get() = lifetime.isExpired

            override fun updateCache() = db.runInTransaction {
                remote.queryAuthors(group) {
                    db.addAuthor(it)
                    db.addGroupAuthor(group, it)
                }
                lifetime.update()
            }

            override fun queryFromCache() = db.queryAuthors(group, visitor)
        })

    override fun queryTracks(author: Author, visitor: Catalog.TracksVisitor) =
        executor.executeQuery("tracks", object : QueryCommand {
            private val lifetime = db.getAuthorTracksLifetime(author, TRACKS_TTL)
            override val isCacheExpired
                get() = lifetime.isExpired

            override fun updateCache() = db.runInTransaction {
                remote.queryTracks(author) {
                    db.addTrack(it)
                    db.addAuthorTrack(author, it)
                }
                lifetime.update()
            }


            override fun queryFromCache() = db.queryTracks(author, visitor)
        })

    override fun queryPictures(author: Author, visitor: Catalog.PicturesVisitor) = executor
        .executeQuery("pictures", object : QueryCommand {
            private val lifetime = db.getAuthorPicturesLifetime(author, PICTURES_TTL)
            override val isCacheExpired
                get() = lifetime.isExpired

            override fun updateCache() = db.runInTransaction {
                remote.queryPictures(author) {
                    db.addAuthorPicture(author, it)
                }
                lifetime.update()
            }

            override fun queryFromCache() = db.queryPictures(author, visitor)
        })

    override fun findTracks(query: String, visitor: Catalog.FoundTracksVisitor) =
        if (remote.searchSupported()) {
            remote.findTracks(query, visitor)
        } else {
            db.findTracks(query, visitor)
        }
}
